src/slot_provider/clock.rs

```diff
@@ -1,88 +1,180 @@
 //! Clock based SlotProvider
-//! When a fn does not finish within a slot (12 seconds), the next call is made concurrently, up to a buffer limit. When the buffer limit is reached, the passed fn will not be called for a new slot, but wait in a FIFO queue until a previous slot has finished.
-use ethers::providers::StreamExt;
```

Comment:
Another general observation:
I agree that these changes reduce type noise, however I think they introduce additional complexity that might outweigh those benefits:

1. Having to maintain our own queue of slots
2. Manual counter / iteration limit implementation ( although that could probably be fixed easily by going back to `take` see discussion above)
3. "Running" two streams instead of one plus additional overhead of communicating between them with streams.

Note that this PR introduces ca. 300 net lines of code. (Although part of that is better testing etc. and it is an imperfect measure of code complexity of course).

How do you think about those ? 
Are there additional benefits of this approach beyond the type noise reduction ? 
Do you have any hypothesis / experience / data on how these changes will affect performance ?

Reply:
We can get rid of the queue of slots, you're right. I made a proposal in the thread above which makes it obsolete.

What I like in the PR and think we can keep, is the idea that launching attestation happens in a thread where everything needed gets passed in wrapped in arcs without having to pass a complex boxed closure.

No clue about performance.